src/main/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluator.java

```diff
@@ -0,0 +1,132 @@
+    private ConfigModel configModel;
+    private DynamicConfigModel dcm;
+    private final AtomicReference<NamedXContentRegistry> namedXContentRegistry;
+
+    public RestLayerPrivilegesEvaluator(
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        AuditLog auditLog,
+        final ClusterInfoHolder clusterInfoHolder,
+        AtomicReference<NamedXContentRegistry> namedXContentRegistry
+    ) {
+
+        super();
+        this.clusterService = clusterService;
+        this.auditLog = auditLog;
+
+        this.threadContext = threadPool.getThreadContext();
+
+        this.clusterInfoHolder = clusterInfoHolder;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    @Subscribe
+    public void onConfigModelChanged(ConfigModel configModel) {
+        this.configModel = configModel;
+    }
+
+    @Subscribe
+    public void onDynamicConfigModelChanged(DynamicConfigModel dcm) {
+        this.dcm = dcm;
+    }
+
+    private SecurityRoles getSecurityRoles(Set<String> roles) {
+        return configModel.getSecurityRoles().filter(roles);
+    }
+
+    public boolean isInitialized() {
+        return configModel != null && configModel.getSecurityRoles() != null && dcm != null;
+    }
+
+    public PrivilegesEvaluatorResponse evaluate(final User user, Set<String> action0) {
+
+        if (!isInitialized()) {
+            throw new OpenSearchSecurityException("OpenSearch Security is not initialized.");
+        }
+
+        final PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();
+
+        final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+
+        Set<String> mappedRoles = mapRoles(user, caller);
+
+        presponse.resolvedSecurityRoles.addAll(mappedRoles);
+        final SecurityRoles securityRoles = getSecurityRoles(mappedRoles);
+
+        final boolean isDebugEnabled = log.isDebugEnabled();
+        if (isDebugEnabled) {
+            log.debug("Evaluate permissions for {} on {}", user, clusterService.localNode().getName());
+            log.debug("Action: {}", action0);
+            log.debug("Mapped roles: {}", mappedRoles.toString());
+        }
+
+        for (String action : action0) {
+            if (!securityRoles.impliesClusterPermissionPermission(action) && !securityRoles.impliesLegacyPermission(action)) {
+                presponse.missingPrivileges.add(action);
+                presponse.allowed = false;
+                log.info(
+                    "No permission match for {} [Action [{}]] [RolesChecked {}]. No permissions for {}",
+                    user,
+                    action,
+                    securityRoles.getRoleNames(),
+                    presponse.missingPrivileges
+                );
+            } else {
+                if (isDebugEnabled) {
+                    log.debug("Allowed because we have ext permissions for {}", action0);
```

Comment:
nit: should we change the wording "ext"/"extension" here?

Reply:
I was thinking about removing 'ext' completely. wdyt?

Reply:
Ya I think that makes 100% sense.